server/src/main/java/org/eclipse/openvsx/mirror/MirrorExtensionService.java

```diff
@@ -0,0 +1,211 @@
+/** ******************************************************************************
+ * Copyright (c) 2022 Precies. Software Ltd and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ * ****************************************************************************** */
+package org.eclipse.openvsx.mirror;
+
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URLConnection;
+import java.nio.channels.Channels;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.eclipse.openvsx.ExtensionService;
+import org.eclipse.openvsx.UserService;
+import org.eclipse.openvsx.entities.UserData;
+import org.eclipse.openvsx.json.ExtensionJson;
+import org.eclipse.openvsx.json.UserJson;
+import org.eclipse.openvsx.repositories.RepositoryService;
+import org.eclipse.openvsx.util.NotFoundException;
+import org.eclipse.openvsx.util.TargetPlatform;
+import org.eclipse.openvsx.util.TimeUtil;
+import org.eclipse.openvsx.util.VersionAlias;
+import org.jobrunr.jobs.context.JobContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestTemplate;
+
+@Component
+public class MirrorExtensionService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MirrorExtensionService.class);
+
+    @Autowired(required = false)
+    DataMirrorService data;
+
+    @Autowired
+    RepositoryService repositories;
+
+    @Autowired
+    RestTemplate backgroundNonRedirectingRestTemplate;
+
+    @Autowired
+    UserService users;
+
+    @Autowired
+    ExtensionService extensions;
+
+    /**
+     * It applies delta from previous execution.
+     */
+    public void mirrorExtension(String namespaceName, String extensionName, UserData mirrorUser, LocalDate lastModified, JobContext jobContext) {
+        var latest = data.getMirror().getExtension(namespaceName, extensionName, null);
```

Comment:
Please move this to line 96. It's ineffecient to get `latest` before you actually need it.